src/exo/range_analysis.py

```diff
@@ -1,104 +1,171 @@
+        constant bound
     """
-    if isinstance(expr, int):
-        return (expr, expr)
 
-    def merge_add(lhs_range, rhs_range):
-        new_lhs = None
-        new_rhs = None
-        if lhs_range[0] is not None and rhs_range[0] is not None:
-            new_lhs = lhs_range[0] + rhs_range[0]
-        if lhs_range[1] is not None and rhs_range[1] is not None:
-            new_rhs = lhs_range[1] + rhs_range[1]
-        return (new_lhs, new_rhs)
-
-    def merge_sub(lhs_range, rhs_range):
-        new_lhs = None
-        new_rhs = None
-        if lhs_range[0] is not None and rhs_range[1] is not None:
-            new_lhs = lhs_range[0] - rhs_range[1]
-        if lhs_range[1] is not None and rhs_range[0] is not None:
-            new_rhs = lhs_range[1] - rhs_range[0]
-        return (new_lhs, new_rhs)
-
-    def merge_mul(lhs_range, rhs_range):
-        # We make sure numbers aren't negative here,
-        # there is probably a way to come up with a correct
-        # range even when the range contains negative numbers
-        if (lhs_range[0] is not None and lhs_range[0] < 0) or (
-            rhs_range[0] is not None and rhs_range[0] < 0
-        ):
-            return (None, None)
-
-        new_lhs = None
-        new_rhs = None
-        if lhs_range[0] is not None and rhs_range[0] is not None:
-            new_lhs = lhs_range[0] * rhs_range[0]
-        if lhs_range[1] is not None and rhs_range[1] is not None:
-            new_rhs = lhs_range[1] * rhs_range[1]
-        return (new_lhs, new_rhs)
-
-    def merge_div(lhs_range, rhs_range):
-        assert isinstance(rhs_range[0], int)
-        assert isinstance(rhs_range[1], int)
-        assert rhs_range[0] == rhs_range[1]
-        assert rhs_range[0] > 0
-
-        if lhs_range[0] is None or lhs_range[0] < 0:
-            return (None, None)
+    base: Optional[LoopIR.expr]
+    lo: Optional[int]
+    hi: Optional[int]
+
+    def get_bounds(self):
+        """
+        Returns a tuple of the form (base + lo, base + hi + 1). If either
+        endpoint is unboundable, returns inf or -inf.
+        """
+        if self.base is None:
+            lo = "-inf" if self.lo is None else str(self.lo)
+            hi = "inf" if self.hi is None else str(self.hi)
+        else:
+            lo = "-inf" if self.lo is None else f"{self.base} + {self.lo}"
+            hi = "inf" if self.hi is None else f"{self.base} + {self.hi + 1}"
+        return lo, hi
 
-        d = rhs_range[0]
+    def get_size(self):
+        if self.lo is None or self.hi is None:
+            return None
+        # +1 because bounds are inclusive
+        return self.hi - self.lo + 1
+
+    def __str__(self):
+        base = "0" if self.base is None else str(self.base)
+        lo = "-inf" if self.lo is None else str(self.lo)
+        hi = "inf" if self.hi is None else str(self.hi)
+        return f"({base}, {lo}, {hi})"
+
+    def __add__(self, other: int | IndexRange) -> IndexRange:
```

Comment:
Can we add assert to check the types of all these operators since some are expecting IndexRange, some are expecting integers, and some accept either? Or would these types hint add any sort of runtime checks, I am not familiar with that feature.